src/main/java/seedu/address/model/reminder/UniqueReminderList.java

```diff
@@ -0,0 +1,139 @@
+
+import java.time.LocalDate;
+import java.util.Iterator;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import javafx.collections.FXCollections;
+import javafx.collections.ObservableList;
+import seedu.address.model.Model;
+import seedu.address.model.person.Person;
+
+/**
+ * A list of Reminders that enforces uniqueness between its elements and does not allow nulls.
+ * A Reminder is considered unique by comparing using {@code Reminder#equals(Reminder)}. This will allow the reminder
+ * to be added or deleted to maintain the uniqueness of the ReminderList
+ * Supports a minimal set of list operations.
+ *
+ */
+public class UniqueReminderList implements Iterable<Reminder> {
+
+    private boolean isReminderListDirty = false;
+
+    private final Model model;
+    private final ObservableList<Reminder> internalList;
+    private final ObservableList<Reminder> internalUnmodifiableList;
+
+    /**
+     * Constructor of UniqueReminderList
+     */
+    public UniqueReminderList(Model model) {
+        this.model = model;
+        internalList = FXCollections.observableArrayList();
+        internalUnmodifiableList = FXCollections.unmodifiableObservableList(internalList);
+    }
+
+    /**
+     * Sets the reminder list to be dirty.
+     * To be called when the reminder list is modified.
+     */
+    public void setReminderListDirty() {
+        isReminderListDirty = true;
+    }
+
+    /**
+     * Returns true if the list contains no Reminders.
+     */
+    public boolean isEmpty() {
+        return internalList.isEmpty();
+    }
+
+    /**
+     * Returns true if the list contains an equivalent Reminder as the given argument.
+     */
+    public boolean contains(Reminder toCheck) {
+        requireNonNull(toCheck);
+        return internalList.stream().anyMatch(toCheck::equals);
+    }
+
+    /**
+     * Returns the backing list as an unmodifiable {@code ObservableList}.
+     */
+    public ObservableList<Reminder> asUnmodifiableObservableList() {
+        return internalUnmodifiableList;
+    }
+
+    /**
+     * Returns the earliest reminder time in the {@code UniqueReminderList}
+     */
+    public long getEarliestReminderTime() {
+        if (internalList.isEmpty()) {
+            return -1;
+        }
+        return internalList.stream().mapToLong(Reminder::getDueTime).min().getAsLong();
+    }
+
+    /**
+     * Returns the list of reminders associated with a specific date.
+     *
+     * @param date The date that which is used to retrieve the list of reminders.
+     * @return the list of reminders mapped from the given date.
+     */
+    public ObservableList<Reminder> getRemindersAfterDate(LocalDate date) {
+        ObservableList<Reminder> reminderList = FXCollections.observableArrayList();
+        List<Reminder> retrievedReminders = internalList.stream()
+                .filter(a -> a.isAfter(date))
+                .collect(Collectors.toList());
+
+        if (retrievedReminders == null) {
```

Comment:
Do we need this? `retrievedReminders` will never be `null` right?

Reply:
Defensive programming, but also it can be?

Reply:
It's collecting the stream into a `List`, so it will just be a `List` right, even if the stream is empty?